packages/context/src/interception-proxy.ts

```diff
@@ -5,25 +5,28 @@
 
 import {Context} from './context';
 import {InvocationArgs, invokeMethodWithInterceptors} from './interceptor';
+import {ValueOrPromise} from './value-promise';
 
 /**
  * Create the Promise type for `T`. If `T` extends `Promise`, the type is `T`,
- * otherwise the type is `Promise<T>`.
+ * otherwise the type is `ValueOrPromise<T>`.
  */
-export type AsPromise<T> = T extends Promise<unknown> ? T : Promise<T>;
+export type AsValueOrPromise<T> = T extends Promise<unknown>
+  ? T
+  : ValueOrPromise<T>;
 
 /**
- * The async variant of a function to always return Promise<R>. If T is not a
- * function, the type is `T`.
+ * The async variant of a function to always return `ValueOrPromise<R>`. If `T`
+ * is not a function, the type is `T`.
  */
 // tslint:disable-next-line:no-unused (possible tslint bug to treat `R` as unused)
 export type AsAsyncFunction<T> = T extends (...args: InvocationArgs) => infer R
```

Comment:
The name is not very accurate - the function is not always async, it can return the value too. I am proposing to rename this type to `AsInterceptedFunction` (describe the intent, not the implementation details).